client/src/main/java/com/aliyun/emr/rss/client/read/DfsPartitionReader.java

```diff
@@ -0,0 +1,173 @@
+  private final int chunkSize;
+  private final int maxInFlight;
+  private final LinkedBlockingQueue<ByteBuf> results;
+  private final AtomicReference<IOException> exception = new AtomicReference<>();
+  private volatile boolean closed = false;
+  private Thread fetchThread;
+  private long offset;
+  private long lastPos;
+  private final FSDataInputStream hdfsInputStream;
+
+  public DfsPartitionReader(RssConf conf, String shuffleKey, PartitionLocation location,
+    TransportClientFactory clientFactory, int startMapIndex, int endMapIndex) throws IOException {
+    chunkSize = (int) RssConf.chunkSize(conf);
+    maxInFlight = RssConf.fetchChunkMaxReqsInFlight(conf);
+    results = new LinkedBlockingQueue<>();
+
+
+    long length = -1;
+    if (endMapIndex != Integer.MAX_VALUE) {
+      // send rpc to the worker and get stream handler which has 12 bytes,
+      // and I think there is no single shuffle file larger than 2^48(256).
+      // So just store offset and length into stream handler.
+      long timeoutMs = RssConf.fetchChunkTimeoutMs(conf);
+      try {
+        TransportClient client = clientFactory.createClient(
+          location.getHost(), location.getFetchPort());
+        OpenStream openBlocks = new OpenStream(shuffleKey, location.getFileName(),
+          startMapIndex, endMapIndex);
+        ByteBuffer response = client.sendRpcSync(openBlocks.toByteBuffer(), timeoutMs);
+        StreamHandle streamHandle = (StreamHandle) Message.decode(response);
+        offset = (long) Utils.convertStreamHandlerToOffsetAndLength(streamHandle.streamId,
+          streamHandle.numChunks)._1;
+        length = (long) Utils.convertStreamHandlerToOffsetAndLength(streamHandle.streamId,
+          streamHandle.numChunks)._2;
+      } catch (IOException | InterruptedException e) {
+        throw new IOException("read shuffle file from hdfs failed, filePath: " +
+                                location.getStorageInfo().getFilePath(), e);
+      }
+      hdfsInputStream = ShuffleClient.getHdfsFs(conf).open(
+        new Path(Utils.getSortedFilePath(location.getStorageInfo().getFilePath())));
+    } else {
+      offset = 0;
+      hdfsInputStream = ShuffleClient.getHdfsFs(conf).open(
+        new Path(location.getStorageInfo().getFilePath()));
+      length = ShuffleClient.getHdfsFs(conf)
+                 .getFileStatus(new Path(location.getStorageInfo().getFilePath())).getLen();
+    }
+    if (offset != 0) {
+      hdfsInputStream.seek(offset);
+    }
+    if (length != 0) {
+      lastPos = offset + length;
+      fetchThread = new Thread(() -> {
+        ByteBuffer buffer = ByteBuffer.allocate(chunkSize);
+        try {
+          byte[] header = new byte[16];
+          while (!closed && offset < lastPos) {
+            Arrays.fill(header, (byte) 0);
+            while (results.size() >= maxInFlight) {
+              Thread.sleep(50);
+            }
+            hdfsInputStream.readFully(offset, header);
```

Comment:
I think return chunkOffsets is better because we can avoid parse each batch then. We can extend StreamHandle to optionally return chunk offsets.